01)	Business problem

We have a company called House Rocket that has a platform where people can sell or buy properties.
The House Rocket CEO wants to maximize profit by closing good deals. The idea is to buy properties that are cheap to sell at a bigger price later. The problem today is to know with property compensates to buy and reform to sell better later. 
Today the company uses the internet and other sources (ask people, research the market, etc.) to find good business. What does the company want from the DC? That the DC analyses the dataset that the company has and finds good opportunities in it. 
What he wants to know:
1-	How many properties are available to buy?
2-	How many characteristics do the properties possess? 
3-	Which are those characteristics? (Number of bedrooms, number of garages, m², site, etc.).
4-	Which is the most expansive property? (The property with the bigger price)
5-	Which is the property with the bigger number of bedrooms?
6-	What is the total of bedrooms in the dataset?
7-	How many houses have two bathrooms?
8-	What is the average price of all houses in the dataset?
9-	What is the average price of homes with two bathrooms?
10-	What is the minimum price for a 3 bedrooms house?
11-	How many houses have more than 300 square meters of living room?
12-	How many houses have more than 2 floors?
13-	How many houses have a sea view?
14-	Of the houses with a sea view, how many have 3 bedrooms?
15-	Of the houses with more than 300 square meters of living room, how many have more than 2 bathrooms?

02)	Planning the solution 

2.1 Planning the final product
- The output: Text with the answers.
- How I will construct the output: Questions | Answers

2.2 Planning the process
- Where is the information? (Excel, Dataset, manual upload, etc.)
Link: https://www.kaggle.com/harlfoxem/housesalesprediction

- How I would do the data collection? (SQL, Python, Streamlit, etc.)
Download the data

-How I will find the answer to the questions?
->How many properties are available to buy? Count the number of lines in the dataset 
-> How many characteristics do the properties possess? Count the number of columns in the dataset
->Which are those characteristics? (Number of rooms, number of garages, m², site, etc.). Show the name of each column
->Which is the most expansive property? (The property with the bigger price) Order the lines by the column price
->Which is the property with the bigger number of rooms? Order the lines by the column number of rooms

2.3 Planning the tools that will be used
-Which tools do I have available? Excel, Python, R
 -Which tool I will use? Python


In [1]:
from typing import Any

import numpy as np
import pandas as pd

In [2]:
# Loading the dataset
data = pd.read_csv ( 'datasets/kc_house_data.csv')

In [3]:
#1-	How many properties are available to buy? - strategy: select the id column and count the number of unique values
len ( data ['id'].unique())
#other possibility
len (data ['id'].drop_duplicates())

21436

In [4]:
#2-	How many characteristics do the properties possess? - strategy: assumptions 'id' and 'date' are not characteristics
len (data.drop( ['id' , 'date'], axis=1 ).columns)

19

In [5]:
#3-	Which are those characteristics? (Number of bedrooms, number of garages, m², site, etc.) - strategy: assumptions
#'id' and 'date' are not characteristics
data.drop( ['id' , 'date'], axis=1 ).columns.tolist()

['price',
 'bedrooms',
 'bathrooms',
 'sqft_living',
 'sqft_lot',
 'floors',
 'waterfront',
 'view',
 'condition',
 'grade',
 'sqft_above',
 'sqft_basement',
 'yr_built',
 'yr_renovated',
 'zipcode',
 'lat',
 'long',
 'sqft_living15',
 'sqft_lot15']

In [6]:
#4-	 Which is the most expansive property? (The property with the bigger price)
data[['id', 'price']].sort_values('price', ascending=False).reset_index(drop=True).loc[0, 'id']

6762700020

In [7]:
#5-	Which is the property with the bigger number of bedrooms?
data[['id', 'bedrooms']].sort_values('bedrooms', ascending=False).reset_index(drop=True).loc[0, 'id']

2402100895

In [8]:
#6-	What is the total of bedrooms in the dataset?
data ['bedrooms'].sum()

72854

In [9]:
#7-	How many houses have two bathrooms? strategy: filter only the 'id' where the column 'bathroom' is equal to 2
houses =len ( data.loc[data['bathrooms'] == 2, [ 'id', 'bathrooms']])
print ('The number of houses with 2 bathrooms are {}'.format( houses ))

The number of houses with 2 bathrooms are 1930


In [10]:
#8- What is the average price of all houses in the dataset? strategy: check if price is a float and not an object
#than take the mean with a round with two decimal places after the point
np.round( data['price'].mean(), 2 )

540088.14

In [11]:
#9-	What is the average price of homes with two bathrooms?
np.round( data.loc[data ['bathrooms' ] == 2, 'price'].mean(), 2 )

457889.72

In [12]:
#10-What is the minimum price for a 3 bedrooms house?
print(data.loc[data['bedrooms'] == 3, 'price'].min())
#other possibility
data.loc[data['bedrooms'] == 3, ['id', 'price']].sort_values('price', ascending=True).head()

82000.0


id    price
8274   3883800011  82000.0
13756  1788900230  86500.0
5866   9320900420  89000.0
18939  4239400300  90000.0
5723   1788800630  96500.0

In [13]:
#11-How many houses have more than 300 square meters of living room? - strategy: firts covert sqft to square meters and then 
#calculate the condition
data['m2'] = data['sqft_living'] * 0.093
len( data.loc[data['m2'] > 300, 'id'])

2258

In [15]:
#12-How many houses have more than 2 floors?
len (data.loc[data ['floors'] > 2, 'id'])


782

In [19]:
#13-How many houses have a sea view?
len( data.loc[data['waterfront'] == 1, 'id'])

163

In [23]:
#14-Of the houses with a sea view, how many have 3 bedrooms?
len ( data.loc[( data['bedrooms'] == 3) & (data['waterfront'] == 1), 'id'])

64

In [ ]:
#15-Of the houses with more than 300 square meters of living room, how many have more than 2 bathrooms?

In [24]:
len ( data.loc[( data ['m2'] > 300 ) & (data[ 'bathrooms'] > 2), 'id'])

2201

1-	How many properties are available to buy? | 21613
2-	How many characteristics do the properties possess? | 19
3-	Which are those characteristics? (Number of bedrooms, number of garages, m², site, etc.). | 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long', 'sqft_living15', 'sqft_lot15'
4-	 Which is the most expansive property? (The property with the bigger price)| 6762700020
5-	Which is the property with the bigger number of bedrooms? | 2402100895 
6-	What is the total of bedrooms in the dataset? | 72854
7-	How many houses have two bathrooms? |1930
8-	What is the average price of all houses in the dataset? | 540088,14
9-	What is the average price of homes with two bathrooms? | 457889.72
10-	What is the minimum price for a 3 bedrooms house? | 82000.0
11-	How many houses have more than 300 square meters of living room? | 2258
12-	How many houses have more than 2 floors? | 782
13-	How many houses have a sea view? | 163
14-	Of the houses with a sea view, how many have 3 bedrooms? | 64
15-	Of the houses with more than 300 square meters of living room, how many have more than 2 bathrooms? | 2201
